tools/btmidi-server.c

```diff
@@ -0,0 +1,1032 @@
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_SONGPOS);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_SONGSEL);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_TUNE_REQUEST);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_CLOCK);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_START);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_CONTINUE);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_STOP);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_SENSING);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_RESET);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_CONTROL14);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_NONREGPARAM);
+	snd_seq_client_info_event_filter_add(info, SND_SEQ_EVENT_REGPARAM);
+
+	err = snd_seq_set_client_info(server->seq_handle, info);
+	if (err < 0) {
+		fprintf(stderr, "Could not set client info: %s (%d)\n", snd_strerror(err), err);
+		goto _err_port;
+	}
+	
+	err = snd_seq_connect_from(server->seq_handle, 0, client, port);
+	if (err < 0) {
+		fprintf(stderr, "Cannot connect from port %d:%d - %s", client, port, snd_strerror(err));
+		goto _err_port;
+	}
+	
+	/* Input file descriptors */
+	snd_seq_poll_descriptors(server->seq_handle, &pfd, 1, POLLIN);
+
+	server->io = io_new(pfd.fd);
+	if (!server->io) {
+		fprintf(stderr, "Could not allocate I/O eventloop\n");
+		goto _err_port;
+	}
+
+	io_set_read_handler(server->io, midi_notify_cb, server, NULL);
+
+	/* Init the MIDI parser */
+	err = midi_read_init(&server->midi_in);
+	if (err < 0) {
+		fprintf(stderr, "Could not initialise MIDI input parser\n");
+		goto _err_port;
+	}
+
+	err = midi_write_init(&server->midi_out, bt_att_get_mtu(server->att) - 3);
+	if (err < 0) {
+		fprintf(stderr, "Could not initialise MIDI output parser\n");
+		goto _err_midi;
+	}
+
+	/* Add MIDI Service */
+	bt_string_to_uuid(&uuid, MIDI_UUID);
+	service = gatt_db_add_service(server->db, &uuid, true, 6);
+
+	/* Add MIDI IO Characteristics */
+	bt_string_to_uuid(&uuid, MIDI_IO_UUID);
+	server->midi_io = gatt_db_service_add_characteristic(service, &uuid,
+						BT_ATT_PERM_READ | BT_ATT_PERM_WRITE,
+						BT_GATT_CHRC_PROP_WRITE_WITHOUT_RESP | BT_GATT_CHRC_PROP_READ | BT_GATT_CHRC_PROP_NOTIFY,
+						NULL, midi_write_cb, server);
+	server->midi_io_handle = gatt_db_attribute_get_handle(server->midi_io);
+
+	/* Add MIDI CCC */
+	bt_uuid16_create(&uuid, GATT_CLIENT_CHARAC_CFG_UUID);
+	gatt_db_service_add_descriptor(service, &uuid,
+					BT_ATT_PERM_READ | BT_ATT_PERM_WRITE,
+					midi_ccc_read_cb, midi_ccc_write_cb, server);
+
+	/* Activate Service */
+	gatt_db_service_set_active(service, true);
+
+	return 0;
+
+_err_midi:
+	midi_read_free(&server->midi_in);
+
+_err_port:
+	snd_seq_delete_simple_port(server->seq_handle, server->seq_port_id);
+
+_err_handle:
+	snd_seq_close(server->seq_handle);
+	server->seq_handle = NULL;
+
+	return -1;
+}
+
+static int advertise(int dev_id)
```

Comment:
I can't call mgmt_new_default() because the function mgmt_new_default/mgmt_new/io_new/mainloop_add_fd/epoll_ctl() returns -1. I have no idea why.